### ISSUE TITLE
segfault when running high dimensional input file (46 columns x 1955 rows)

### ISSUE THREAD

**Issue by LMHmedchem** [tomilov/quickhull#4]
Hello,

The folks at the opensuse forum have been helping me to build this app in the hope of running it on this high dimensional data.

This is the thread at opensuse
https://forums.opensuse.org/showthread.php/519164-need-cmake-3-1-or-higher-in-opensuse-13-2-yast-only-has-3-0

As you can see, there were some opensuse configuration issues to address before getting the app built. There are segfault events from both quickhull and qh.

From quickhull

(gdb) run < PCA_46-inputs.txt
Starting program: /usr/bin/quickhull < PCA_46-inputs.txt
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
input file: stdin
dimensionality of input is 46
rbox command line:
input points count = 1955
epsilon = 0
simplex time = 423152us

Program received signal SIGSEGV, Segmentation fault.
0x000000000041e2c5 in std::__1::__split_buffer<quick_hull<std::__1::__forward_list_iterator<std::__1::__forward_list_node<std::__1::forward_list<double, std::__1::allocator<double> >, void_>_>, double>::facet_, std::__1::allocator<quick_hull<std::__1::__forward_list_iterator<std::__1::__forward_list_node<std::__1::forward_list<double, std::__1::allocator<double> >, void_>_>, double>::facet_> >::push_back(quick_hull<std::__1::__forward_list_iterator<std::__1::__forward_list_node<std::__1::forward_list<double, std::__1::allocator<double> >, void_>_>, double>::facet*&&) (this=0x7fffffffdd50, 
        __x=<unknown type in /usr/lib/debug/usr/bin/quickhull.debug, CU 0x15b, DIE 0x54748>) at /usr/include/c++/v1/__split_buffer:565
    565        if (__end_ == __end_cap())

and from qh

(gdb) run < PCA_46-inputs.txt
Starting program: /usr/bin/qh < PCA_46-inputs.txt
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
dimensionality of input is 46
rbox command line:
input points count = 1955
epsilon = 2.22045e-16
simplex time = 416578us

Program received signal SIGSEGV, Segmentation fault.
0x000000000042dd8f in std::__1::__hash_table<unsigned long,
std::__1::hash<unsigned long>, std::__1::equal_to<unsigned long>,
std::__1::allocator<unsigned long> >::__construct_node (this=<error
reading variable: Cannot access memory at address 0x7fffff7fecf8>,
__v=<error reading variable: Cannot access memory at address
0x7fffff7fecf0>, __hash=<error reading variable: Cannot access memory
0x7fffff7fecf0>at address 0x7fffff7fece8>)
0x7fffff7fecf0>at /usr/include/c++/v1/__hash_table:2088 2088

Sorry for not using code tags on the above output but I can't seem to get github to format my message correctly with the code tage inserted.

There appears to be a trap for the number of input dimensions being beyond the supported range, so this would appear to be a bug. I have attached the input file here in case that is of use. The thread at opensuse has the output  from running cmake. Please let me know if there is anything else you  need to know.

This was built on opensuse 13.2 x86_64 and I believe that others at opensuse have tried with Leap 42.1 and Tumbleweed.

Thanks.
[PCA_46-inputs.txt.zip](https://github.com/tomilov/quickhull/files/394184/PCA_46-inputs.txt.zip)

**Comment by tomilov** [comment]
@LMHmedchem  I sure it is almost impossible to solve D=46 N=1955 problem on modern computers due to memory limitations. Output growth combinatorically when dimensionality growth linearly. Even for D=10 and N=30 I get a huge amounts of memory consumed during algorithm running. Another problem is epsilon. For D=20+ you need precision better, then **double** or **long double** can give you.

**Comment by LMHmedchem** [comment]
Thanks you for your comments. The machine I am using has 16GB of memory and it is not coming close to running out at the segfault. The issue must be related to the precision or some other issue you suggest. I am surprised that the number of rows is all that relevant.

Are you familiar with a method for computing the chull in high dimensional space? I have read some about D-optimal designs and a method for calculating the chull based on the determinant of the square of the input matrix and its transpose. I haven't been able to get that method to give expected results even in 2D. 

I am also looking at a leave one out method that seems like it could work.

**Comment by tomilov** [comment]
@LMHmedchem how long qh works until it crashed?

![screenshot from 2016-08-08 07-31-32](https://cloud.githubusercontent.com/assets/896175/17483356/45c0808a-5db8-11e6-8139-595e5c4329e6.png)

I tried to wait until RelWithDebInfo crashed - it is about 20 minutes and place, where cause of the error origins, is memory allocation somewhere in STL. 1 GB of memory footprint at that moment is reached.

I used yesterday's version of `clang` and `libc++` from trunk (clang 4.0 rev. 277825). Total memory available on _Ubuntu 16.04_ on virtual machine is about 3.7 GB. Memory consumption growth very slowly during programm running.
